### DISCUSSION TITLE
Adding an NTPServer to ClockService

### DISCUSSION THREAD

**Discussion by fdizazzo** [eclipse-kura/kura#5142] [Ideas]
Hi all,
just got yet another 'frontier' use case: having devices who can only talk to a Kura-equipped edge gateway but could find a centralized clock syncronization beneficial.
So, could adding a NTP Server into ClockService (or somewhere else into the codebase) be useful/something you're interested in?

After a quick overview of the as-is:
- No additional dependencies are required;
- It would work best with the java-ntp provider, fine with the others;
- It would be disabled by default, for some ease of mind about security I'd leave to a user the creation of the firewall rule. 

What do you think? :)

**Comment by mattdibi** [comment]
Hello there @fdizazzo 
I briefly discussed your proposal with @MMaiero and we have couple observations to make.

1. I don't think you strictly need to implement something inside Kura (event though every contribution is welcome 😉) to achieve what you want: spinning up a container with an NTP server inside would probably suffice in your case (excluding performance limitation consideration of course). See [cturra/docker-ntp](https://github.com/cturra/docker-ntp).
2. We don't think the `ClockService` is a good fit for what you want to achieve: it is meant to be an NTP **client** (and already quite big in scope due to how many underlying technologies it's leveraging). Adding an NTP **server** functionality there would be out-of-scope for that particular service. 
3. A better way to achieve this would be to create a new standalone component (something called "NTPServerService"). In this case you don't even need to contribute directly to the Kura codebase but can have [a separate bundle](https://eclipse.github.io/kura/docs-release-5.4/java-application-development/configurable-application/#add-dependencies-to-manifest) and distribute it through the Eclipse Marketplace (like we do).

Let me know what you think or if you need any clarification 😄

**Reply by fdizazzo** [reply]
Hi all,
first of all thank your for your time 😜 
I pretty much agree with all your points, the ntp server can be a separate component.

My main need for this would have been being able to determine how 'good' my system time is and provide the information to the NTP Server, namely the Stratum and last syncronization time. 
Now this can also be achieved with minor changes in the ClockService to retrieve then via events and/or apis.

On an unrelated note, have we ever thought about being able to use a GPS to have straight-from-satellite clock synchronization? 🤔

**Reply by mattdibi** [reply]
> On an unrelated note, have we ever thought about being able to use a GPS to have straight-from-satellite clock synchronization? 🤔

Yup we can leverage the NTPD service for doing so, see [API Javadocs](https://download.eclipse.org/kura/docs/api/5.4.0/apidocs/). I'm not aware of reference implementations though 🤔  I would need to ask more senior team members for more details.

![image](https://github.com/eclipse/kura/assets/22748355/ac79daf2-46e0-4890-bef4-3c6ba21850ba)

**Reply by mattdibi** [reply]
@fdizazzo At today's Kura Dev meeting I was made aware of the [APIs exposed by the PositionService](https://github.com/eclipse/kura/blob/develop/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java) interface that allows for retrieving of the date/time through GPS.

For an implementation you can take a look at: https://github.com/eclipse/kura/tree/develop/kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position

**Reply by fdizazzo** [reply]
Thanks! I'll double check if my use case can be implemented with the existing APIs , and/or share a draft PR if not. Sadly no ETA on that 🤔 

Also, I think the shared calendar for DevMeetings went out of sync after the holiday break, pointing at the wrong weeks :)

**Reply by mattdibi** [reply]
> Also, I think the shared calendar for DevMeetings went out of sync after the holiday break, pointing at the wrong weeks :)

That would explain the low attendance :D. I'll see what I can do, thanks for reporting the issue :)

**Reply by mattdibi** [reply]
@fdizazzo [We have updated the post with the calendar](https://github.com/eclipse/kura/discussions/4841#discussioncomment-8663410). Feel free to ping us if you find any problem.

**Comment by mattdibi** [comment]
Closing this. Feel free to re-open this if needed.
